include/jemalloc/internal/jemalloc_internal_inlines_c.h

```diff
@@ -52,56 +68,81 @@ isalloc(tsdn_t *tsdn, const void *ptr) {
 }
 
 JEMALLOC_ALWAYS_INLINE void *
-iallocztm(tsdn_t *tsdn, size_t size, szind_t ind, bool zero, tcache_t *tcache,
-    bool is_internal, arena_t *arena, bool slow_path) {
+iallocztm_explicit_slab(tsdn_t *tsdn, size_t size, szind_t ind, bool zero,
+    bool slab, tcache_t *tcache, bool is_internal, arena_t *arena,
+    bool slow_path) {
 	void *ret;
 
+	assert(!slab || sz_can_use_slab(sz_s2u(size))); /* slab && large is illegal */
 	assert(!is_internal || tcache == NULL);
 	assert(!is_internal || arena == NULL || arena_is_auto(arena));
 	if (!tsdn_null(tsdn) && tsd_reentrancy_level_get(tsdn_tsd(tsdn)) == 0) {
 		witness_assert_depth_to_rank(tsdn_witness_tsdp_get(tsdn),
 		    WITNESS_RANK_CORE, 0);
 	}
 
-	ret = arena_malloc(tsdn, arena, size, ind, zero, tcache, slow_path);
+	ret = arena_malloc(tsdn, arena, size, ind, zero, slab, tcache, slow_path);
 	if (config_stats && is_internal && likely(ret != NULL)) {
 		arena_internal_add(iaalloc(tsdn, ret), isalloc(tsdn, ret));
 	}
 	return ret;
 }
 
+JEMALLOC_ALWAYS_INLINE void *
+iallocztm(tsdn_t *tsdn, size_t size, szind_t ind, bool zero, tcache_t *tcache,
+    bool is_internal, arena_t *arena, bool slow_path) {
+	bool slab = sz_can_use_slab(sz_s2u(size));
```

Comment:
Can we avoid the `sz_s2u` here? If we don't that, which paths are failing the assert inside `sz_can_use_slab`?

Reply:
At least in the test-cases, we fail [coming from `prof_tdata_init_impl`](https://github.com/jemalloc/jemalloc/blob/90176f8a87a0b5bdb0ac4c1a515b1d9c58dc5a82/src/prof_data.c#L1137) in `test/integration/aligned_alloc` and we also fail in `a0ialloc` in `test_a0`. There might be more beyond this that we don't know about.

Reply:
I'm a bit worried the extra s2u here is going to result in added instructions or less optimized code. Try adding the s2u to these call sites? If it's only a few of those, I'd say do that to make sure no regression on this path.

Reply:
I just realized, this is the non-alignment case. Let's simply remove the s2u? We'd lose the assertion inside sz_can_use_slab, but that's no different from the current logic, i.e. using a non-usize to compare against SMALL_MAXCLASS.

Reply:
Good point. I've removed the call here and the assertion.